### PR TITLE
[Ready to be merged] Add RNN-LM to Conformer-CTC decoding

### PR DESCRIPTION
**Pull request by ezerhouni**
Continuation of: https://github.com/k2-fsa/icefall/pull/127

### CODE REVIEW

**Comment by ezerhouni** [comment]
@csukuangfj I refactored some RNN-LM code and moved it to `icefall`. For now I just moved the `ptb` egs and didn't test it. Should I try to fix/test it or remove it and keep only the librispeech.

File: egs/librispeech/ASR/RESULTS.md [hunk 1]
```diff
@@ -1139,17 +1139,18 @@ You can find the tensorboard log at: <https://tensorboard.dev/experiment/D7NQc3x
 
 #### 2021-11-09
 
-The best WER, as of 2021-11-09, for the librispeech test dataset is below
-(using HLG decoding + n-gram LM rescoring + attention decoder rescoring):
+The best WER, as of 2022-20-06, for the librispeech test dataset is below
```
[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md]: Fixed

File: egs/librispeech/ASR/RESULTS.md [hunk 2]
```diff
@@ -1139,17 +1139,18 @@ You can find the tensorboard log at: <https://tensorboard.dev/experiment/D7NQc3x
...
+(using HLG decoding + n-gram LM rescoring + attention decoder rescoring + rnn lm rescoring):
 
 |     | test-clean | test-other |
 |-----|------------|------------|
-| WER | 2.42       | 5.73       |
+| WER | 2.32       | 5.39       |
```
[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1147]: Yes. I tried several parameters for training (RNN) and all of them improved the current setup. I pushed the one that worked the best. Couple of things to note :
1. In quite some experiment, `ngram_lm_scale` was almost 0 (like 0.01 or 0.05), and attention_lm_scale and rnn_lm_scale being roughly similar. Hence, I wonder if it worth to do an n-gram rescoring before.
2. Testing is very slow (on GPU), hence, I was thinking to maybe add decoding on multi-GPUs (to accelerate the process).

**Comment by ezerhouni** [comment]
@csukuangfj I removed the `rnn_lm` folder from ptb. I think this PR is ready.

File: egs/librispeech/ASR/RESULTS.md [hunk 3]
```diff
@@ -1170,11 +1171,27 @@ export CUDA_VISIBLE_DEVICES="0,1,2,3"
   --start-epoch 0 \
   --num-epochs 90
 # Note: It trains for 90 epochs, but the best WER is at epoch-77.pt
+
+# Train the RNN-LM
```
Comment (csukuangfj) [inline egs/librispeech/ASR/RESULTS.md:1175]: Could you upload the trained RNN-LM model and the training logs to huggingface and place a link to it here? 
By the way, what is the PPL on test-clean of this RNN-LM model?

File: egs/librispeech/ASR/RESULTS.md [hunk 4]
```diff
@@ -1170,11 +1171,27 @@ export CUDA_VISIBLE_DEVICES="0,1,2,3"
...
+cd icefall
```
[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1176]: Since it is code that is useful to most of the egs, I was thinking to follow what is found in kaldi, (script/rnnlm), so that it is easily accessible

[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/librispeech/ASR/RESULTS.md:1176]: ok, it is fine with me.

File: egs/librispeech/ASR/RESULTS.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1147]: It is done by computing n-best scores: [code](https://github.com/k2-fsa/icefall/pull/439/files#diff-39edd6cc0561e6565939f4d4f2213b6d5f13ebb8128b933685dcffb892939db6R1038)
I was also suprised of how slow it was. Finally, I tried several parameters (num_layers and embedding_dim), and while I could see a difference of speed during training, it was barely noticeable at inference time. I will try to have a look. (FYI, I am testing on V100 32gb)

File: egs/ptb/LM/local/prepare_lm_training_data.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline egs/ptb/LM/local/prepare_lm_training_data.py:1]: Shall we replace it with a symlink to `egs/librispeech/ASR/local/prepare_lm_training_data.py` ?

File: egs/ptb/LM/local/train_bpe_model.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline egs/ptb/LM/local/train_bpe_model.py:1]: Also, can we replace it with a symlink to `egs/librispeech/ASR/local/train_bpe_model.py`?

**Review by csukuangfj — COMMENTED** [review]
Left some minor comments.
Thanks!

File: egs/librispeech/ASR/RESULTS.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1147]: The batch_size might be relatively small actually (see part of the log for test-clean)
```
2022-06-19 03:13:02,731 INFO [decode.py:584] batch 0/?, cuts processed until now is 2
/opt/conda/lib/python3.8/site-packages/lhotse/dataset/sampling/dynamic.py:329: UserWarning: The first cut drawn in batch collection violates the max_frames, max_cuts, or max_duration constraints - we'll return it anyway. Consider increasing max_frames/max_cuts/max_duration.
  warnings.warn(
2022-06-19 03:27:19,920 INFO [decode.py:584] batch 100/?, cuts processed until now is 331
2022-06-19 03:42:01,574 INFO [decode.py:584] batch 200/?, cuts processed until now is 719
2022-06-19 03:58:27,017 INFO [decode.py:584] batch 300/?, cuts processed until now is 1128
2022-06-19 04:15:55,729 INFO [decode.py:584] batch 400/?, cuts processed until now is 1492
2022-06-19 04:30:15,283 INFO [decode.py:584] batch 500/?, cuts processed until now is 1827
2022-06-19 04:45:39,058 INFO [decode.py:584] batch 600/?, cuts processed until now is 2217
2022-06-19 05:00:28,432 INFO [decode.py:584] batch 700/?, cuts processed until now is 2479
```

File: egs/librispeech/ASR/RESULTS.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1176]: Closing it

**Comment by csukuangfj** [comment]
> @csukuangfj I removed the `rnn_lm` folder from ptb. I think this PR is ready.

Could you please upload the pre-trained RNN LM model to huggingface?
The steps are below:
1. Go to https://huggingface.co/ and register an account. (You may use your github account for direct login)
2. Create a new repo, e.g., `icefall-librispeech-rnn-lm`
3. Install `git-lfs`. If you are using Ubuntu, it can be installed with `sudo apt-get install git-lfs`.
    Refer to https://git-lfs.github.com/ for installation on other operating systems.
4. Clone the repo you just created using the following command:
```bash
git lfs install
git clone https://huggingface.co/<your_user_name_on_hugging_face>/icefall-librispeech-rnn-lm
cd icefall-librispeech-rnn-lm
// then copy the averaged RNN LM model you are using for rescoring to here.
//
// You can refer to export.py in the librispeech recipe for how to export the model

git add <your_expoerted_model.pt>
git commit -m "some message"
git push
# Enter your username and password for authentication.
# You may need to enter the above information twice.

# You may place a readme.md in the repo and add a link in it to this PR.

```

File: egs/ptb/LM/local/train_bpe_model.py [hunk 7]
```diff
@@ -0,0 +1 @@
+/Users/ezerhoun/repos/open_source/icefall/egs/librispeech/ASR/local/train_bpe_model.py
```
[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/ptb/LM/local/train_bpe_model.py]: My bad, fixed it

File: egs/librispeech/ASR/RESULTS.md (hunk 3 above)
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1175]: The ppl on test-clean and test-other (from prepare.sh) is 9.970. Do you want test-clean in particular ?

Comment (danpovey) [reply egs/librispeech/ASR/RESULTS.md:1175]: Hm... we'll have to look in papers, how people report the LM perplexity.  But normally the perplexity would be per word, not per BPE token.  I.e. you would have to take 9.970 ** { num_words_in_test_set / num_bpe_tokens_in_test_set}

**Comment by ezerhouni** [comment]
@csukuangfj I uploaded the model and update the code according to reviews

File: egs/librispeech/ASR/RESULTS.md (hunk 3 above)
Comment (danpovey) [reply egs/librispeech/ASR/RESULTS.md:1175]: BTW, we plan to work on an un-normalized version of RNN-T that will benefit much more, I think, from improved language models.  So the value of this language modeling stuff is more than just the immediate WER gains.  We need to gain some experience with training neural LMs.

**Comment by csukuangfj** [comment]
> @csukuangfj I uploaded the model and update the code according to reviews

Thanks!

File: egs/librispeech/ASR/RESULTS.md (hunk 3 above)
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1175]: There are 211138 tokens and 104919 words in test set, so the perplexity per word should be 3.13.

The un-normalized version of RNN-T sounds pretty interesting !

Comment (danpovey) [reply egs/librispeech/ASR/RESULTS.md:1175]: I had the formulat the wrong way around, should be ** {num_tokens/num_words}

File: egs/librispeech/ASR/RESULTS.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (XhrLeokk) [reply egs/librispeech/ASR/RESULTS.md:1147]: hello there ezerhouni, I'm not sure do u still remember this proj...

 I have trained my own RNNLM as icefall doc recommanded(3layer LSTM, dim is 2048, after 30epochs PPL on vali is 9.87).

Then I used this LM on "shallowfusion decode method" (add lm_score as beam searching), but I can't even get one positive result compair to the original method "modified beam search". (which can get WER 3.11 on clean, 7.94 on other)

No matter what value the lambda is. (as the experiments shows, larger the LM lambda, the worse result I get, example: with 0.29, i got   4.63 and 10.43)

confused, did I miss something Important? or any other suggests..

[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1147]: @XhrLeokk Have you tried negative value ?

[the remark this answers is not in the document]
Comment (XhrLeokk) [reply egs/librispeech/ASR/RESULTS.md:1147]: yes I tested -0.09 once, but doesn't help much..
LM
these are the exps I did with my RNNLM.

however, now I have started another training loop with --use_fp16 0 to do more epochs to see if anything changes

[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1147]: @XhrLeokk I also had a hard time training an RNN-LM on my own dataset. Maybe you could try with an attention based one (I haven't tried).

[the remark this answers is not in the document]
Comment (XhrLeokk) [reply egs/librispeech/ASR/RESULTS.md:1147]: @ezerhouni kind a relief to know... After recur this RNNLM I will go try TFMLM, if got some breakthrough I'll post the results here. :)

[the remark this answers is not in the document]
Comment (ezerhouni) [reply egs/librispeech/ASR/RESULTS.md:1147]: @XhrLeokk Yes, please let me know how it went :)
